// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package messengerServer;

service MessengerServer {
  rpc Communicate (stream ClientMessage) returns (stream ServerMessage) {}
}

message Subscribe {
  string tag = 1;
}

message OutputMessage {
  bytes receiver = 1;
  string tag = 2;
  bytes content = 3;
}

message ClientMessage {
  oneof client_message {
    Subscribe subscribe = 1;
    OutputMessage output_message = 2;
  }
}

message InputMessage {
  string tag = 2;
  bytes content = 3;
}

message ServerMessage {
  oneof server_message {
    InputMessage input_message = 1;
  }
}
